testonly/integration/clientserver.go

```diff
@@ -91,87 +99,145 @@ func getTestDB(testID string) (*sql.DB, error) {
+	var env LogEnv
+	var err error
+	env.DB, env.dbURI, err = getTestDB(testID)
 	if err != nil {
 		return nil, err
 	}
 
 	timesource := &util.SystemTimeSource{}
-	registry, err := builtin.NewDefaultExtensionRegistry()
+	registry, err := builtin.NewExtensionRegistry(env.dbURI)
 	if err != nil {
 		return nil, err
 	}
 
-	// Start Log Server.
-	grpcServer := grpc.NewServer()
-	logServer := server.NewTrillianLogRPCServer(registry, timesource)
-	trillian.RegisterTrillianLogServer(grpcServer, logServer)
+	// Create Log Server.
+	env.grpcServer = grpc.NewServer()
+	env.logServer = server.NewTrillianLogRPCServer(registry, timesource)
+	trillian.RegisterTrillianLogServer(env.grpcServer, env.logServer)
 
-	// Start Sequencer.
+	// Create Sequencer.
 	keyManager, err := crypto.LoadPasswordProtectedPrivateKey(privateKeyFile, privateKeyPassword)
 	if err != nil {
 		return nil, err
 	}
-	sequencerManager := server.NewSequencerManager(keyManager, registry,
+	env.LogOperation = server.NewSequencerManager(keyManager, registry,
 		sequencerWindow)
-	sequencerTask := server.NewLogOperationManagerForTest(ctx, registry,
-		batchSize, sleep, timeSource, sequencerManager)
+	if numSequencers == 0 {
+		// Test sequencer that needs manual triggering (with env.Sequencer.OperationLoop()).
+		env.Sequencer = server.NewLogOperationManagerForTest(ctx, registry,
+			batchSize, sleepBetweenRuns, timeSource, env.LogOperation)
+	} else {
+		// Start a live sequencer in a goroutine.
+		ctx, env.sequencerCancel = context.WithCancel(ctx)
+		env.Sequencer = server.NewLogOperationManager(ctx, registry,
+			batchSize, numSequencers, sleepBetweenRuns, timeSource, env.LogOperation)
+		env.pendingTasks.Add(1)
+		go func(env *LogEnv) {
+			defer env.pendingTasks.Done()
+			env.Sequencer.OperationLoop()
+		}(&env)
+	}
 
 	// Listen and start server.
 	addr, lis, err := listen()
 	if err != nil {
 		return nil, err
 	}
-	go grpcServer.Serve(lis)
+	env.pendingTasks.Add(1)
+	go func(env *LogEnv, lis net.Listener) {
+		defer env.pendingTasks.Done()
+		env.grpcServer.Serve(lis)
+	}(&env, lis)
 
 	// Connect to the server.
-	cc, err := grpc.Dial(addr, grpc.WithInsecure())
+	env.ClientConn, err = grpc.Dial(addr, grpc.WithInsecure())
 	if err != nil {
 		return nil, err
 	}
 
-	return &LogEnv{
-		grpcServer:   grpcServer,
-		logServer:    logServer,
-		ClientConn:   cc,
-		DB:           db,
-		LogOperation: sequencerManager,
-		Sequencer:    sequencerTask,
-	}, nil
+	return &env, nil
+}
+
+// LaunchCT adds a running CT personality to the LogEnv test environment.
+func (env *LogEnv) LaunchCT(cfgs []ct.LogConfig) error {
+	var err error
+
+	// Provision the logs.
+	for _, cfg := range cfgs {
+		if err := env.CreateLog(cfg.LogID); err != nil {
+			return fmt.Errorf("failed to provision log %d: %v", cfg.LogID, err)
+		}
+	}
+	// Wait long enough for a signer run to happen, so empty tree gets signed.
+	time.Sleep(2 * sleepBetweenRuns)
```

Comment:
Also, does `OS.Signal` do anything clever with threads/goroutines?  Because UNIX signals and threads traditionally don't mix well...